contracts/StableJoeStaking.sol

```diff
@@ -0,0 +1,312 @@
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/SafeERC20Upgradeable.sol";
+
+/**
+ * @title Stable JOE Staking
+ * @author Trader Joe
+ * @notice StableJoeStaking is a contract that allows JOE deposits and receives stablecoins sent by JoeMakerV4's daily
+ * harvests. Users deposit JOE and receive a share of what has been sent by JoeMakerV4 based on their participation of
+ * the total deposited JOE
+ * Every time `updateReward(token)` is called, We distribute the balance of that tokens as rewards to users that are
+ * currently staking inside this contract.
+ */
+contract StableJoeStaking is Initializable, OwnableUpgradeable {
+    using SafeMathUpgradeable for uint256;
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    /// @notice Info of each user
+    struct UserInfo {
+        /// @notice How many LP rewardTokens the user has provided
+        uint256 amount;
+        /// @notice The sum of token rewards already paid to this account. See explanation below
+        mapping(IERC20Upgradeable => uint256) rewardDebt;
+        /**
+         * @notice We do some fancy math here. Basically, any point in time, the amount of JOEs
+         * entitled to a user but is pending to be distributed is:
+         *
+         *   pending reward = (user.amount * accRewardPerShare) - user.rewardDebt[token]
+         *
+         * Whenever a user deposits or withdraws JOE. Here's what happens:
+         *   1. accRewardPerShare (and `lastRewardBalance`) gets updated
+         *   2. User receives the pending reward sent to his/her address
+         *   3. User's `amount` gets updated
+         *   4. User's `rewardDebt[token]` gets updated
+         */
+    }
+
+    IERC20Upgradeable public joe;
+
+    /// @notice Internal balance of $JOE, this gets updated on user deposits / withdrawals
+    /// @dev this allows to reward users with $JOE.
+    uint256 internalJoeBalance;
+    /// @notice The array of token that people can claim to get reward
+    IERC20Upgradeable[] public rewardTokens;
+    mapping(IERC20Upgradeable => bool) public isRewardToken;
+    /// @notice Last reward balance of `token`
+    mapping(IERC20Upgradeable => uint256) public lastRewardBalance;
+
+    address public feeCollector;
+
+    /// @notice The deposit fee, scaled to `PRECISION`
+    uint256 public depositFeePercent;
+
+    /// @notice Accumulated `token` rewards per share, scaled to `PRECISION`. See above
+    mapping(IERC20Upgradeable => uint256) public accRewardPerShare;
+    /// @notice `PRECISION` of `accRewardPerShare`
+    uint256 public PRECISION;
+
+    /// @notice Info of each user that stakes JOE
+    mapping(address => UserInfo) private userInfo;
+
+    /// @notice Emitted when a user deposits JOE
+    event Deposit(address indexed user, uint256 amount);
+
+    /// @notice Emitted when a user withdraws JOE
+    event Withdraw(address indexed user, uint256 amount);
```

Comment:
Should we also have a `ClaimReward` event or add `claim_token_amount` into the `Withdraw` event?

Reply:
I don't know, but yeah we could.
@cryptofish7 do you have any preference ?